packages/flutter_tools/lib/src/compute_dev_dependencies.dart

```diff
@@ -2,126 +2,136 @@
+/// In case of `flutter run --no-pub` this will at most ignore new dependencies.
+Map<String, Dependency> computeTransitiveDependencies(
+  FlutterProject project,
+  PackageConfig packageConfig,
+  FileSystem fileSystem, {
+  bool followDevDependencies = false,
+}) {
+  final Map<String, Dependency> result = <String, Dependency>{};
+
+  final List<String> packageNamesToVisit = <String>[project.manifest.appName];
+  while (packageNamesToVisit.isNotEmpty) {
+    final String current = packageNamesToVisit.removeLast();
+    if (result.containsKey(current)) {
+      continue;
     }
-    if (value is! List<Object?>) {
-      fail('Expected field "$name" to be a list, got "$value"');
+    final YamlNode pubspec;
+    final List<String> dependencies = <String>[];
+    final Uri rootUri;
+    if (current == project.manifest.appName) {
+      rootUri = project.directory.uri;
+      pubspec = project.manifest.toYaml();
+      dependencies.addAll(project.manifest.dependencies);
+    } else {
+      final Package? package = packageConfig[current];
+      if (package == null) {
+        continue;
+      }
+      rootUri = package.root;
+      if (rootUri.scheme != 'file') {
+        continue;
+      }
+
+      final String pubspecPath = fileSystem.path.fromUri(rootUri.resolve('pubspec.yaml'));
+      try {
+        pubspec = loadYamlNode(
+          fileSystem.file(pubspecPath).readAsStringSync(),
+          sourceUrl: Uri.file(pubspecPath),
+        );
+      } on IOException {
+        continue;
+      } on FormatException {
+        continue;
+      }
+      if (pubspec is! YamlMap) {
+        continue;
+      }
+      final dynamic dependenciesMap = pubspec['dependencies'] ?? <String, String>{};
+      if (dependenciesMap is! Map) {
+        continue;
+      }
+
+      for (final dynamic name in dependenciesMap.keys) {
+        if (name is! String) {
+          continue;
+        }
+        dependencies.add(name);
+      }
     }
-    return <T>[
-      for (final Object? any in value)
-        if (any is T) any else fail('Expected element to be a $T, got "$any"'),
-    ];
-  }
-
-  // Parse the JSON roughly in the following format:
-  //
-  // ```json
-  // {
-  //   "root": "my_app",
-  //   "packages": [
-  //     {
-  //       "name": "my_app",
-  //       "kind": "root",
-  //       "dependencies": [
-  //         "foo_plugin",
-  //         "bar_plugin"
-  //       ],
-  //       "directDependencies": [
-  //         "foo_plugin"
-  //       ],
-  //       "devDependencies": [
-  //         "bar_plugin"
-  //       ]
-  //     }
-  //   ]
-  // }
-  // ```
-  final List<Map<String, Object?>> packages = asListOrFail(jsonResult['packages'], 'packages');
-
-  Map<String, Object?> packageWhere(
-    bool Function(Map<String, Object?>) test, {
-    required String reason,
-  }) {
-    return packages.firstWhere(test, orElse: () => fail(reason));
+    packageNamesToVisit.addAll(dependencies);
+    if (current == project.manifest.appName) {
```

Comment:
This will only happen during the first iteration of the loop, right? If so, can we hoist it out of the loop?